content/blog/2019-10-23-loop-unrolling-with-bril.md

```diff
@@ -0,0 +1,154 @@
+
+#### Static Analysis
+The first part of this project is to implement a completely static compiler pass to do loop unrolling for Bril programs. 
+Our pass performs complete unrolling only. 
+This pass calls several passes that are already present in the Bril repo. In addition, we implemented a reaching definition
+pass using the code template in ``examples/df.py``, updated the constant propagation pass to make it more powerful,
+and slightly changed the cfg construction pass  in ``examples/cfg.py`` so that each basic block has a label. The overall 
+flow of the pass is as follows:
+- Construct the CFG, then compute predecessors, successors, and dominators of each basic block. 
+- Run the constant propagation and reaching definition passes. 
+- Use CFG information to find loops in the program. Notice that we only unroll loops with regular structures. Our definition
+of "regular structure" is given in the implementation detail section. 
+- For each loop that can be unrolled, leverage the information from constant propagation and reaching definition passes to calculate the tripcount. 
+- If the tripcount is computable, decide whether we want to unroll the loop or not using a simple criterion: the total 
+number of instructions in the loop (tripcount * number of instructions in all loop basic blocks). Currently we unroll loops
+with less than 1k instructions. 
+- Unroll the loop by first duplicating all basic blocks inside the loop and then removing redundant control flow instructions. 
+
+#### Dynamic Profiling
+
+The key idea of dynamic profiling is to estimate the loop status by simulating the loop execution. Since the loop trip count may depend on different conditions, it will hard to infer the loop bound without the information from actual execution environment. For example, induction variable being udpated indirectly can make the unrolling problem complicated. An intuitive way of simplifying the problem is to proflie the loops before unrolling. Here we extend the data flow analysis pass to support dynamic profiling, i.e. we maintain a status table to track the update history for each variables and . By refering to the table and locate the update history of the back-edge related instructions and basic blocks, we can easily infer various information needed, e.g. trip count, induction variable as well as loop stride. Here to simplify the analysis, we assume all loops to be unrolled are natural loops, namely the program can only enter the loop from the header. Under the assumption that the loop bound is invariant in multiple runs, the basic blocks in the loop body will be repliacted for trip bound times. Back edges (i.e. jmps) and condition check statements will be removed form originla bril program.
```

Comment:
I can't quite tell from this discussion whether this dynamically-informed loop unrolling is sound—that is, does the pass insert exit checks?

Put differently, if you observe the loop executing 30 times, will you unroll it 30 times? So the program will be incorrect if run on a different input that needs the loop to execute 29 times? If not, it might be a good idea to clarify why this is not the case.